File: src-py/bc.py
import datetime
from hashlib import sha256
import json
class Transaction:
    def __init__(self, sender, recipient, amount):
        self.sender = sender
        self.recipient = recipient
        self.amount = amount
        self.timestamp = datetime.datetime.now()


    def print_transaction(self):
        print("From:", self.sender)
        print("To:", self.recipient)
        print("Amount:", self.amount)

    def generate_hash(self):
        transaction_contents = str(self.sender) + str(self.recipient) + str(self.amount)
        transaction_hash = sha256(transaction_contents.encode())
        return transaction_hash.hexdigest()



class Block:
    def __init__(self,minerAddress, timestamp, transactions, previous_hash):
        self.minerAddress = minerAddress
        self.timestamp = timestamp
        self.transactions = transactions
        self.previous_hash = previous_hash
        self.nonce = 0
        self.hash = self.hash_block()

    def hash_block(self):
        return sha256(str(str(self.timestamp) + str(self.transactions) + str(self.previous_hash) +str(self.nonce)).encode()).hexdigest()

    

    #proof of work algorithm
    def mine_block(self, difficulty):
        while self.hash[:difficulty] != "0"*difficulty:
            self.nonce += 1
            self.hash = self.hash_block()

        print("Block mined:", self.hash)



    def print_block(self):
        print("==========================================================")
        print("timestamp:", self.timestamp)
        block = {
            "minerAddress": self.minerAddress,
            "previous_hash": self.previous_hash,
            "nonce": self.nonce,
            "hash": self.hash,
            "transactions": []
        }
        
        for transaction in self.transactions:
            block["transactions"].append({

                "sender": transaction.sender,
                "recipient": transaction.recipient,
                "amount": transaction.amount
            })

        print(json.dumps(block, indent=4))

        ##printing all transactions



class Blockchain:
    def __init__(self):
        self.chain = []
        self.transactions = []
        self.difficulty = 3
        self.create_genesis_block()

    def create_genesis_block(self):
        self.chain.append(Block("SYSTEM",datetime.datetime.now(), [], "0"))

    def get_last_block(self):
        return self.chain[-1]
    
    def add_block(self, block):
        self.chain.append(block)
    
    def add_transaction(self, transaction):
        self.transactions.append(transaction)
    
    def mine_block(self, minerAddress):
        last_block = self.get_last_block()
        new_block = Block(minerAddress,datetime.datetime.now(), self.transactions, last_block.hash)
        new_block.mine_block(self.difficulty)
        self.chain.append(new_block)
        self.transactions = []


        return new_block
    
    def print_blockchain(self):
        for block in self.chain:
            block.print_block()
    
    def is_chain_valid(self):
        for i in range(1, len(self.chain)):
            current_block = self.chain[i]
            previous_block = self.chain[i-1]
            if current_block.hash != current_block.hash_block():
                return False
            if current_block.previous_hash != previous_block.hash_block():
                return False
        return True

    def get_balance(self, address):
        balance = 0
        for block in self.chain:
            for transaction in block.transactions:
                if transaction.sender == address:
                    balance -= transaction.amount
                if transaction.recipient == address:
                    balance += transaction.amount
        return balance

    def get_transactions(self, address):
        transactions = []
        for block in self.chain:
            for transaction in block.transactions:
                if transaction.sender == address or transaction.recipient == address:
                    transactions.append(transaction)
        return transactions

class Miner:
    def __init__(self, id, name, power, header, private_key, public_key):
        self.id = id
        self.name = name
        self.power = power
        self.header = header
        self.private_key = private_key
        self.public_key = public_key


def main():
    import random
    miners = list()

    for i in range(0, 10):
        miners.append(Miner(i, "Miner"+str(i), 100, "Header", "Private Key", "Public Key"))

    blockchain = Blockchain()
    
    

    
    for miner in miners:
        from_miner = random.randint(0,len(miners))
        to_miner = random.randint(0,len(miners))
        if(from_miner == to_miner):
            to_miner = random.randint(0,len(miners))
        amount = random.randint(1,100)
        blockchain.add_transaction(Transaction(from_miner, to_miner,amount))
        blockchain.mine_block(miner.id)


    
    blockchain.print_blockchain()


if __name__ == "__main__":
    main()

File: src/bc.js
const SHA256 = require("crypto-js/sha256");
// const _ = require('lodash');
const elliptic = require('elliptic').ec; // elliptic curve
const ec = new elliptic('secp256k1');

// const key = ec.genKeyPair();
// const publicKey = key.getPublic('hex');

//tworzymy  zmienną daty w celu update'u daty co iteracje
let date = new Date(Date.now());

//tworzymy klase transaction która będzie przechowywać dane transakcji oraz czas wykonania transakcji który zostanie wykorzystany do wygenerowania hasha transakcji
class Transaction {
    constructor(fromAddress, toAddress, amount) {
        this.fromAddress = fromAddress;
        this.toAddress = toAddress;
        this.amount = amount;
        this.timestamp = date;
    }

    //metoda hashująca transakcję
    calculateHash() {
        return SHA256(this.fromAddress + this.toAddress + this.amount + this.timestamp).toString();
    };
    
    //metoda podpisująca transakcję
    signTransaction(signingKey) {
        if (signingKey.getPublic('hex') !== this.fromAddress){
            console.log('Cannot sign transaction with another address');
            return;
        }
        const transactionHash = this.calculateHash();
        const signature = signingKey.sign(transactionHash, 'base64');
        this.signature = signature.toDER('hex');
    }

    isTransactionValid(){
        if(this.fromAddress === "SYSTEM") return true;
        if(!this.signature || this.signature.length === 0){
            console.log('No signature in this transaction');
            return;
        }

        const publicKey = ec.keyFromPublic(this.fromAddress, 'hex');
        return publicKey.verify(this.calculateHash(), this.signature);
    }

}

class Mesh {

    constructor(id) {
        this.id = id;
        this.peerList = [];
    }

    addToMesh(miner) {
        this.peerList.push(miner);
    }

    showMesh() {
        console.log("Peer List: ");
        for (var i = 0; i < this.peerList.length; i++) {
            console.log(this.peerList[i].id);
        }
    }

    checkIfExsits(miner) {
        for (var i = 0; i < this.peerList.length; i++) {
            if (this.peerList[i].id === miner.id) {
                return true;
            }
        }
        return false;
    }

    removePeer(miner) {
        for (var i = 0; i < this.peerList.length; i++) {
            if (this.peerList[i].id === miner.id) {
                this.peerList.splice(i, 1);
                console.log("Miner " + miner.id + " has been removed from the mesh network");
            }
        }
    }

    getIndexForMiner(){
        if(this.peerList.length === 0) return -1;
        return this.peerList.length;
    }

}

class Miner {
    constructor(id, name, power, address) {
        this.id = id;
        this.name = name;
        this.power = power;
        this.currHeader = null;
        this.privateAddress = ec.genKeyPair().getPrivate('hex'); //prywatny
        this.publicAddress = ec.keyFromPrivate(this.privateAddress).getPublic('hex'); 
    }
}

class Block {
    constructor(minerAddress, timestamp, transactions, previousHash = '') {

        this.minerAddress = minerAddress;
        this.timestamp = timestamp;
        this.previousHash = previousHash;
        this.nonce = 0;
        this.hash = this.calculateHash();
        this.transactions = transactions;
    }

    calculateHash() {
        return SHA256(this.index + this.previousHash + this.timestamp
            + JSON.stringify(this.transactions) + this.nonce).toString();
    }

    proofOfWork(difficulty) {

        while (this.hash.substring(0, difficulty) !== Array(difficulty + 1).join("0")) {
            this.nonce++;
            this.hash = this.calculateHash();
        }
    }

    areTransactionsValid(){
        for(const transaction of this.transactions){
            if(!transaction.isTransactionValid()) return false;
        }
        return true;
    }
}


class Blockchain {

    constructor() {
        this.chain = [this.createGenesisBlock()];
        this.difficulty = 3;
        this.miningReward = 15; // 15 reward seems balanced
        this.pendingTransactions = [];
    }
    //constructor(minerAddress, timestamp, transactions, previousHash = '')

    createGenesisBlock() {
        return new Block("SYSTEM", this.getDate(), [] , "0");
    }

    getDate() {
        if (date !== new Date(Date.now())) date = new Date(Date.now()); 
        return date.getDate()
            + "/" + date.getMonth() + 1
            + "/" + date.getFullYear()
            + " " + date.getHours()
            + ":" + date.getMinutes()
            + ":" + date.getSeconds();
    }

    getLastBlock() {
        return this.chain[this.chain.length - 1];
    }

    broadcastBlock(broadcaster, block, meshNetwork) {
        for (var i = 0; i < meshNetwork.peerList.length; i++) {
            meshNetwork.peerList[i].currHeader = block.hash;
        }
    }

    
    addTransaction(transaction){

        if(!transaction.fromAddress || !transaction.toAddress || !transaction.isTransactionValid()){
            console.log('invalid transaction') 
            return
        }
        // if(!transaction.isTransactionValid()) throw new Error('Invalid transaction');
        if(transaction.amount <= 0){
            console.log('invalid amount')
            return;
        }
        if(this.checkBalance(transaction.fromAddress) < transaction.amount) {
            console.log("Insufficient balance");
            console.log(`Trying to send ${transaction.amount} with ${this.checkBalance(transaction.fromAddress)} in wallet`);
            return;
        }

        this.pendingTransactions.push(transaction);
    }
    
    checkBalance(address){
        //find the balance of the address by looking through the wallets
        let balance = 5; // startowe fundusze
        for(const block of this.chain){
            for(const transaction of block.transactions){

                if(transaction.fromAddress === address) balance -= transaction.amount;
                if(transaction.toAddress === address) balance += transaction.amount;
            }
        }
        return balance;
    }

    mineBlock(address, meshNetwork, miner, howManyIterations) {
        console.log(`Miner ${miner.id} with power ${miner.power} is mining block ${this.chain.length}`);
        return new Promise((resolve, reject) => {
            if (this.chain.length === howManyIterations) {

                console.log("Blockchain: ");
                for (var i = 0; i < this.chain.length; i++) {
                    console.log(this.chain[i]);
                }

                reject();
            }
            const reward = new Transaction("SYSTEM", address, this.miningReward);
            this.pendingTransactions.push(reward);

            const costs = new Transaction(address, "Tauron Energia", (miner.power / this.miningReward) * 0.01 );
            costs.signTransaction(ec.keyFromPrivate(miner.privateAddress));
            this.pendingTransactions.push(costs);

            const block = new Block(address, this.getDate(), this.pendingTransactions, this.getLastBlock().hash);
            
            block.proofOfWork(this.difficulty);
            this.chain.push(block);
            
            this.broadcastBlock(miner, block, meshNetwork)
            this.pendingTransactions = [];
            resolve(block);
        });


    }


    isChainValid() {

        for (let i = 1; i < this.chain.length; i++) {
            const currentBlock = this.chain[i];
            const previousBlock = this.chain[i - 1];
            if (currentBlock.hash !== currentBlock.calculateHash()) return false;
            if (currentBlock.previousHash !== previousBlock.hash) return false;
            if(!currentBlock.areTransactionsValid()) return false;
        }
        return true;
    }

}


module.exports = {

    Transaction,
    Mesh,
    Miner,
    Block,
    Blockchain,
    
}
